decisions/2024-08-30_central_sync_queue_pull_reduction.md

```diff
@@ -0,0 +1,164 @@
+- Fixes the problem for existing sites all versions of OMS and legacy apps.
+
+_Cons:_
+
+- Still potentially queuing lots of records to many sites. Some users will have >1000 sites and a general master list used across them with >5000 items. These instances will still suffer millions of queued records that central_change_log set out to fix to begin with!
+- Eventually we want OMS only deployments where OMS controls central data, so a solution in LCS land would require us eventually fixing it again in OMCS land.
+- Doing work in LCS when we're trying to focus on OMS
+
+### Option 2 - LCS central_change_log filtering per log processed
+
+We implement a system that allows sync v5 change log requests to have some filtering on the LCS side.
+
+Suppose a site requests 5000 change_log records from cursor 0:
+
+1. If the change_log record is for a list_master_line,
+2. Get all the list_master.ID values for lists that are visible to the site; that is from list_master_name_join for all stores (this can be done once for the request and reused in the process, or cached between requests to reduce querying overheads)
+3. Get the list_master_line record
+4. If the list_master_line.master_list_id is **not** in that set, skip this change log record.
+
+_Pros:_
+
+- Stay in changelog land and don't queue loads of records
+
+_Cons:_
+
+- Still looks like 90k records to pull, even if large amounts get filtered out.
+- If 5000 records are requested and we filter out some (or all) of them, do we scan ahead to grab some more records till the batch size is fulfilled (and continuing until done)? Probably negligible but none the less a slower.
+- Added overhead to getting records
+- If they actually need all of the master list lines, it's more overhead doing filtering for no reason. Can probably optimise this case though.
+- Doing work in LCS when we're trying to focus on OMS
+- Eventually we want OMS only deployments where OMS controls central data, so a solution in LCS land would require us eventually fixing it again in OMCS land.
+- Only fixes for OMS
+
+### Option 3 - OMCS change_log filtering per log processed
+
+The filtering could be more or less the same as in Option 2, but we do it in OMCS. Go full hog: We change to OMCS changelog pull and drop pulling LCS changelog.
+
+_Pros:_
+
+- Stay in changelog land and don't queue loads of records
+- Simplify sync progress in OMS, going from 3 steps back to 2 steps (just pull queued from sync v5, pull changelog from sync v6)
+- Future proof solution for the ages
+
+_Cons:_
+
+- Maybe not all records on sync v6 could come through? I don't see why that wouldn't be possible
+- Only fixes for OMS
+
+### Option 4 - OMCS change_log filter based on change_log.parent_id
+
+When master list line records are added to the changelog, we include a `parent_id` (could crudely use store_id for non-store records...)
```

Comment:
This sounds a little bit confusing to me, does every record have a parent id? We need to look up master lists for every record before we can save it?

I think I'd rather do the filtering when we're querying change log...

Reply:
I had a similar idea though. I was thinking we could have a single map with `record_id -> Vec<store_id>`
That would make the filtering quick at filtering time, just check if the requesting store needs this record or not.
It would have the same problem though, where we're need to calculate visibility for every central record when we create or update it.

Reply:
As per my question above I'm starting to think that we might need to have some kind of state management for what's visible at each site when records are saved, and related records might need to be sent (e.g. create a new cursor_id if visiblity changes) This has the challenge that it would potentially create 1000s of out going records to other sites that don't need the update!

Reply:
> This sounds a little bit confusing to me, does every record have a parent id? We need to look up master lists for every record before we can save it?
>

For relevant tables it'd be recorded in the trigger where we already have the FK to put int the `changelog.parent_id`.
 
> I think I'd rather do the filtering when we're querying change log...

I agree but looking for ideas on how to do that effectively. Not sure how to do it without an ID to check. In a nutshell need to check `changelog.master_list_line.master_list_id IN site.stores.master_list_store_join.master_list_id // only for the changelogs for table == master_list_line`